PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/LinkAzureServiceBeforeRunProvider.java

```diff
@@ -0,0 +1,192 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.*;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+public class LinkAzureServiceBeforeRunProvider extends BeforeRunTaskProvider<LinkAzureServiceBeforeRunTask> {
+    private static final Logger LOGGER = Logger.getInstance(SelectSubscriptionsAction.class);
+    private static final String NAME = "Link Azure Service";
+    private static final String DESCRIPTION = "Link Azure Service Task";
+    private static final String SPRING_BOOT_CONFIGURATION_REF = "com.intellij.spring.boot.run.SpringBootApplicationRunConfiguration";
+    public static final Key<LinkAzureServiceBeforeRunTask> ID = Key.create("LinkAzureServiceBeforeRunProviderId");
+    public static final Key<Boolean> LINK_AZURE_SERVICE = Key.create("LinkAzureService");
+    public static final Key<Map<String, String>> LINK_AZURE_SERVICE_ENVS = Key.create("LinkAzureServiceEnvs");
+
+    @Override
+    public Key<LinkAzureServiceBeforeRunTask> getId() {
+        return ID;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public String getDescription(LinkAzureServiceBeforeRunTask task) {
+        return DESCRIPTION;
+    }
+
+    @Nullable
+    @Override
+    public Icon getIcon() {
+        return AzureIconLoader.loadIcon(AzureIconSymbol.Common.AZURE);
+    }
+
+    @Nullable
+    @Override
+    public LinkAzureServiceBeforeRunTask createTask(@NotNull RunConfiguration runConfiguration) {
+        boolean alwaysEnable = false;
+        if (StringUtils.equals(runConfiguration.getClass().getName(), SPRING_BOOT_CONFIGURATION_REF)
+        /*|| runConfiguration instanceof com.intellij.javaee.appServers.run.configuration.CommonStrategy*/) {
+            alwaysEnable = true;
+        }
+        return new LinkAzureServiceBeforeRunTask(getId(), alwaysEnable);
+    }
+
+    @Override
+    public boolean executeTask(@NotNull DataContext dataContext, @NotNull RunConfiguration runConfiguration,
+                               @NotNull ExecutionEnvironment executionEnvironment, @NotNull LinkAzureServiceBeforeRunTask linkAzureServiceBeforeRunTask) {
+        String moduleName = this.getModuleName(runConfiguration);
+        if (StringUtils.isBlank(moduleName)) {
+            return true;
+        }
+        Map<String, String> linkedEnvMap = new LinkedHashMap<>();
+        retrieveEnvMap(runConfiguration.getProject(), linkedEnvMap, moduleName);
+        if (MapUtils.isNotEmpty(linkedEnvMap)) {
+            // set envs for remote deploy
+            if (runConfiguration instanceof WebAppConfiguration) {
+                IntelliJWebAppSettingModel model = ((WebAppConfiguration) runConfiguration).getModel();
+                AzureWebAppMvpModel.getInstance().updateWebAppSettings(model.getSubscriptionId(), model.getWebAppId(), linkedEnvMap, new HashSet<>());
```

Comment:
`WebAppConfiguration` may create new web apps here, so webappId may be null as it is not created yet during before run stages